x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -202,43 +229,83 @@ public Automaton allowedActionsMatcher(String index) {
+                if (indexCheck) {
+                    boolean actionCheck = group.checkAction(action);
+                    // mapping updates are allowed for certain privileges on indices and aliases (but not on data streams),
+                    // outside of the privilege definition
+                    boolean bwcMappingActionCheck = isMappingUpdateAction && false == isDataStream && false == isBackingIndex &&
+                            group.privilege().name().stream().anyMatch(privilegeNameSetBwcAllowMappingUpdate::contains);
+                    if (actionCheck || bwcMappingActionCheck) {
+                        // propagate DLS and FLS permissions over the concrete indices
+                        for (String index : concreteIndices) {
+                            Set<FieldPermissions> fieldPermissions = fieldPermissionsByIndex.computeIfAbsent(index, (k) -> new HashSet<>());
+                            fieldPermissionsByIndex.put(indexOrAlias, fieldPermissions);
+                            fieldPermissions.add(group.getFieldPermissions());
+                            DocumentLevelPermissions permissions =
+                                    roleQueriesByIndex.computeIfAbsent(index, (k) -> new DocumentLevelPermissions());
+                            roleQueriesByIndex.putIfAbsent(indexOrAlias, permissions);
+                            if (group.hasQuery()) {
+                                permissions.addAll(group.getQuery());
+                            } else {
+                                // if more than one permission matches for a concrete index here and if
+                                // a single permission doesn't have a role query then DLS will not be
+                                // applied even when other permissions do have a role query
+                                permissions.setAllowAll(true);
+                            }
+                        }
+                    }
+                    granted |= actionCheck;
+                    bwcGrantMappingUpdate |= bwcMappingActionCheck;
+                    if (false == actionCheck && bwcMappingActionCheck) {
+                        for (String privilegeName : group.privilege.name()) {
+                            if (privilegeNameSetBwcAllowMappingUpdate.contains(privilegeName)) {
+                                bwcDeprecationLogActions.add(() -> {
+                                    deprecationLogger.deprecate("[" + indexOrAlias + "] mapping update for ingest privilege [" +
+                                            privilegeName + "]", "the mapping update action [" + action + "] on the [" +
+                                            indexOrAlias + "] index, is granted by the [" + privilegeName + "] privilege," +
+                                            " but the privilege has been tightened to not allow it in the next major release");
+                                });
+                            }
```

Comment:
> I wonder whether it would be better to display a single deprecation log per indexOrAlias, e.g. ... is granted by [privA, privB] ....

I dunno. I've chosen the current variant because it is easier to code.

The `deprecationLogger.deprecate` method has built-in logic to deduplicate frequent log entries with the same key.
If a log message contains a list of privileges, instead of a single one, the deduplication logic would have to be smarter to compare that "all (or some of) the privileges (given the index) on this deprecation log entry have appeared in (some) previous entries". It sounds too complicated to implement.

Reply:
> Also this if block can be placed inside the previous if (actionCheck || bwcMappingActionCheck) so its own if check can become just if (false == actionCheck).

I can make the move, I don't know of any heuristic that could help me choose between the two.